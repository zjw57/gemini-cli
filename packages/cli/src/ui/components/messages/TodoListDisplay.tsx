/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import type React from 'react';
import { Box, Text } from 'ink';
import type { Todo, TodoList, TodoStatus } from '@google/gemini-cli-core';
import { theme } from '../../semantic-colors.js';

export interface TodoListDisplayProps {
  todos: TodoList;
  terminalWidth: number;
}
const TodoStatusDisplay: React.FC<{ status: TodoStatus }> = ({ status }) => {
  switch (status) {
    case 'completed':
      return <Text color={theme.status.success}>✓</Text>;
    case 'in_progress':
      return <Text color={theme.text.accent}>»</Text>;
    case 'pending':
      return <Text color={theme.text.primary}>☐</Text>;
    case 'cancelled':
      return <Text color={theme.status.error}>✗</Text>;
    default:
      return null;
  }
};

export const TodoListDisplay: React.FC<TodoListDisplayProps> = ({
  todos,
  terminalWidth,
}) => (
  <Box flexDirection="column" width={terminalWidth}>
    {todos.todos.map((todo: Todo, index: number) => (
      <Box key={index} flexDirection="row">
        <Box marginRight={1}>
          <TodoStatusDisplay status={todo.status} />
        </Box>
        <Box flexShrink={1}>
          <Text color={theme.text.primary}>{todo.description}</Text>
        </Box>
      </Box>
    ))}
  </Box>
);
