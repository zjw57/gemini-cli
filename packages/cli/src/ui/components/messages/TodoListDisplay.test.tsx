/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { render } from 'ink-testing-library';
import { describe, it, expect } from 'vitest';
import { TodoListDisplay } from './TodoListDisplay.js';
import type { TodoList, TodoStatus } from '@google/gemini-cli-core';

describe('<TodoListDisplay />', () => {
  const terminalWidth = 80;

  it('renders an empty todo list correctly', () => {
    const todos: TodoList = { todos: [] };
    const { lastFrame } = render(
      <TodoListDisplay todos={todos} terminalWidth={terminalWidth} />,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders a todo list with various statuses correctly', () => {
    const todos: TodoList = {
      todos: [
        { description: 'Task 1', status: 'pending' as TodoStatus },
        { description: 'Task 2', status: 'in_progress' as TodoStatus },
        { description: 'Task 3', status: 'completed' as TodoStatus },
        { description: 'Task 4', status: 'cancelled' as TodoStatus },
      ],
    };
    const { lastFrame } = render(
      <TodoListDisplay todos={todos} terminalWidth={terminalWidth} />,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders a todo list with long descriptions that wrap', () => {
    const todos: TodoList = {
      todos: [
        {
          description:
            'This is a very long description for a pending task that should wrap around multiple lines when the terminal width is constrained.',
          status: 'pending' as TodoStatus,
        },
        {
          description:
            'Another completed task with an equally verbose description to test wrapping behavior.',
          status: 'completed' as TodoStatus,
        },
      ],
    };
    const { lastFrame } = render(
      <TodoListDisplay todos={todos} terminalWidth={40} />,
    );
    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders a single todo item', () => {
    const todos: TodoList = {
      todos: [{ description: 'Single task', status: 'pending' as TodoStatus }],
    };
    const { lastFrame } = render(
      <TodoListDisplay todos={todos} terminalWidth={terminalWidth} />,
    );
    expect(lastFrame()).toMatchSnapshot();
  });
});
