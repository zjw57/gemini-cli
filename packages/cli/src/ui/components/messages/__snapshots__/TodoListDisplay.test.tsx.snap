// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<TodoListDisplay /> > renders a single todo item 1`] = `"☐ Single task"`;

exports[`<TodoListDisplay /> > renders a todo list with long descriptions that wrap 1`] = `
"☐ This is a very long description for a
  pending task that should wrap around
  multiple lines when the terminal width
  is constrained.
✓ Another completed task with an equally
  verbose description to test wrapping
  behavior."
`;

exports[`<TodoListDisplay /> > renders a todo list with various statuses correctly 1`] = `
"☐ Task 1
» Task 2
✓ Task 3
✗ Task 4"
`;

exports[`<TodoListDisplay /> > renders an empty todo list correctly 1`] = `""`;
